Initial IndexedDB Support

arihant2math opened the pull request:
Adds indexeddb support to servo. At the moment heed is being used as the backend, although this can be swapped out by implementing `KvsEngine`. This PR adds a thread + a thread pool for Indexeddb related operations. Also `database_access_task_source` is added for Indexeddb related operations.

This is a partial rewrite of #25214. (Reopened due to branching issue)

Fixes #6963

arihant2math committed:
Add tests for indexeddb and feature flag

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

Co-authored-by: Rasmus Viitanen <rasviitanen@gmail.com>
Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math committed:
Add initial IndexedDB Support

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

Co-authored-by: Rasmus Viitanen <rasviitanen@gmail.com>
Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

jdm commented:
The duplicate package results from. `./mach tidy` can be addressed by updating the ignore list in servo-tidy.toml.

arihant2math commented:
Ah ok, I'll do that

jdm commented:
Looks like `./mach test-unit` needs some compile fixes:
```
error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> components/fonts/tests/font_context.rs:177:5
    |
177 |     ResourceThreads::new(core_sender, storage_sender)
    |     ^^^^^^^^^^^^^^^^^^^^----------------------------- an argument of type `IpcSender<IndexedDBThreadMsg>` is missing
    |
note: associated function defined here
   --> /home/runner/work/servo/servo/components/shared/net/lib.rs:353:12
    |
353 |     pub fn new(
    |            ^^^
help: provide the argument
    |
177 |     ResourceThreads::new(core_sender, storage_sender, /* IpcSender<IndexedDBThreadMsg> */)
    |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

arihant2math commented:
I fixed that error but I'm currently getting a lot of unrelated errors involving the XR web API (specifically the bindgen).

arihant2math commented:
Nothing is broken per-say, but for some reason the pages load way slower

mrobinson commented on components/net/Cargo.toml:
```diff
@@ -47,6 +47,7 @@ percent-encoding = { workspace = true }
 pixels = { path = "../pixels" }
 profile_traits = { workspace = true }
 rayon = { workspace = true }
+rkv = "0.10"
```
You should be able to de-duplicate `syn` by upgrading to the most recent version of `rkv` instead of using this one, which is 5 years old.

arihant2math replied:
Switching to heed instead of rkv, so this should be resolved.

mrobinson commented:
Please squash you changes into a single commit (or organize them somehow, so that they are easier to review) and add @rasviitanen [as a co-author](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) on these changes (since they were the original author). Thank you!

arihant2math committed:
Improved performance and reduced crashes and hangs

Switch idb backend to heed and add support for more key values and used better concurrency measures and updated interface between backend and scripts.

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math commented:
I still can't understand why the actual browser doesn't work, it seems like there isn't anything blocking any thread, but it seems certain threads relating to layout are being blocked in some way.

gterzian commented:
> I still can't understand why it doesn't pass the smoke test. 

I'm guessing it relates to https://github.com/servo/servo/blob/58e27e3708acc94a6cf76e47fa511dd19a69177b/components/net/indexeddb/idb_thread.rs#L109

arihant2math commented:
> > I still can't understand why it doesn't pass the smoke test.
> 
> I'm guessing it relates to
> 
> https://github.com/servo/servo/blob/58e27e3708acc94a6cf76e47fa511dd19a69177b/components/net/indexeddb/idb_thread.rs#L109

I would have assumed this blocks the IDB Threadpool, which is separate from everything else.

arihant2math commented:
Did some further testing, with some odd results. The smoketest passed on an aarch64 Mac.

```
% ./mach smoketest
No build type specified so assuming `--dev`.
No message: Disconnected (thread IndexedDBManager, at components/net/indexeddb/idb_thread.rs:165)
42 threads are still running after shutdown (bad).
```

arihant2math committed:
added ability to close stores and part of the code for counting the database

also redid identifier check

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

jdm commented:
You will need another `./mach fmt`.

arihant2math commented:
Ok it should work now

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

jdm commented:
I am slowly working on reviewing this PR.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math committed:
removed excessive error

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

jdm commented on components/net/indexeddb/idb_thread.rs:
```diff
@@ -0,0 +1,343 @@
+                let mut db_dir = self.idb_base_dir.clone();
+                db_dir.push(&idb_description.as_path());
+                if std::fs::remove_dir_all(&db_dir).is_err() {
+                    sender.send(Err(())).unwrap();
+                } else {
+                    sender.send(Ok(())).unwrap();
```
```suggestion
                    let _ = sender.send(Err(()));
                } else {
                    let _ = sender.send(Ok(()));
```

jdm replied:
Let's make similar changes to the other uses of unwrap/expect on send calls in this file.

jdm commented on components/net/indexeddb/idb_thread.rs:
```diff
@@ -0,0 +1,343 @@
+                    .expect("Could not send commit status");
+            },
+            SyncOperation::UpgradeVersion(sender, origin, db_name, _txn, version) => {
+                self.get_database_mut(origin, db_name).map(|db| {
+                    db.version = version;
+                });
```
```suggestion
                if let Some(db) = self.get_database_mut(origin, db_name) {
                    db.version = version;
                };
```

jdm replied:
And please make similar changes to other uses of map in this file that don't use the return value.

jdm reviewed (changes requested):
I've read all the code! I looked at some specifications while reviewing some of it, but there are enough methods with known missing parts that I mostly focused on code clarity and major concepts. It will be easier to review individual methods for correctness as we implement them more fully in smaller pieces.

jdm commented:
>It will be easier to review individual methods for correctness as we implement them more fully in smaller pieces.

To be clear, I'm suggesting that subsequent work should happen in new PRs; there's no need to make this one any larger in my opinion.

arihant2math committed:
added docstrings

added docstrings

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math committed:
addressed part of the review

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math commented on components/script/dom/webidls/IDBVersionChangeEvent.webidl:
```diff
@@ -0,0 +1,25 @@
+ *
+ */
+
+// FIXME:(arihant2math) Expose to Worker too
```
There were some compilation errors that prevented me from doing that.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math commented on components/net/indexeddb/idb_thread.rs:
```diff
@@ -0,0 +1,343 @@
+                };
+                self.databases.remove(&idb_description);
+
+                // FIXME:(rasviitanen) Possible security issue?
```
Probably using `remove_dir_all` with arbitrary input. I added a note below it with clarification

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

jdm commented:
I haven't been keeping track—is this ready for a new round of reviews at this point?

arihant2math commented:
> I haven't been keeping track—is this ready for a new round of reviews at this point?

Not yet, there are still are few more major issues that need to be resolved.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math commented:
I've returned to this PR. Should be back into shape for another review soon.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

arihant2math committed:
fix rebuild

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math commented on components/net/indexeddb/engines/heed.rs:
```diff
@@ -0,0 +1,269 @@
+                .expect("Failed to open db_dir")
+        };
+        // FIXME:(arihant2math) What is a reasonable number of threads?
+        let threads = 4;
```
@jdm I think it does.
Is there a common way that other components share a threadpool somehow?

jdm replied:
Put an Arc around it! https://github.com/servo/servo/blob/main/components/net/image_cache.rs#L434

arihant2math replied:
Ah ok, I'll do that after I fix the build.

arihant2math commented:
I'm getting:
```
error[E0277]: `*mut ()` cannot be sent between threads safely
   --> components\script\dom\idbopendbrequest.rs:168:13
    |
167 |           global.task_manager().database_access_task_source().queue(
    |                                                               ----- required by a bound introduced by this call
168 | /             task!(send_upgradeneeded_notification: move || {
169 | |                 let this = this.root();
170 | |                 let txn = trusted_transaction.root();
171 | |                 let conn = connection.root();
...   |
211 | |                 this.dispatch_success(&*conn, can_gc);
212 | |             }),
    | |______________^ `*mut ()` cannot be sent between threads safely
    ```
    I can't seem to see why this happens.

jdm commented:
@arihant2math That happens when you try to use a `CanGc` value from outside of a task inside of a task. Use `CanGc::note()` instead.

arihant2math commented:
Thanks, I'm down to this:
```
error[E0277]: `RefCell<std::option::Option<ScriptEventLoopSender>>` cannot be shared between threads safely
   --> components\script\dom\idbrequest.rs:215:13
    |
215 | /             Box::new(move |message| {
216 | |                 let response_listener = response_listener.clone();
217 | |                 task_source.queue(task!(request_callback: move || {
218 | |                     response_listener.handle_async_request_finished(
219 | |                         message.expect("Could not unwrap message"));
220 | |                 }));
221 | |             }),
    | |______________^ `RefCell<std::option::Option<ScriptEventLoopSender>>` cannot be shared between threads safely
    |
    ```
    
and

    ```
    error[E0308]: mismatched types
    --> components\script\dom\idbopendbrequest.rs:96:41
     |
96   |                 open_request.set_result(js::jsapi::UndefinedHandleValue);
     |                              ---------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Handle<'_, Value>`, found `Handle<Value>`    
     |                              |
     |                              arguments to this method are incorrect
     |
     = note: `Handle<Value>` and `Handle<'_, Value>` have similar names, but are actually distinct types
note: `Handle<Value>` is defined in crate `mozjs_sys`
    --> D:\Documents\Programming\GitHub\servo\target\debug\build\mozjs_sys-d189f7a5766231b6\out/build/jsapi.rs:5767:9
     |
5767 |         pub struct Handle<T> {
     |         ^^^^^^^^^^^^^^^^^^^^
note: `Handle<'_, Value>` is defined in crate `mozjs`
    --> C:\Users\ariha\.cargo\git\checkouts\mozjs-dd2a25693c61d87d\728acdf\mozjs\src\gc\root.rs:114:1
     |
114  | pub struct Handle<'a, T: 'a> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: method defined here
    --> components\script\dom\idbopendbrequest.rs:219:12
     |
219  |     pub fn set_result(&self, result: HandleValue) {
     |            ^^^^^^^^^^        -------------------
     ```
     
and

```
error[E0599]: no method named `to_jsval` found for type `usize` in the current scope
   --> components\script\dom\idbobjectstore.rs:242:44
    |
242 | ...                   string_len.to_jsval(*cx, return_val);
    |                                  ^^^^^^^^ method not found in `usize`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
note: `ErrorToJsval` defines an item `to_jsval`, perhaps you need to implement it
   --> components\script\dom\bindings\error.rs:278:1
    |
278 | pub(crate) trait ErrorToJsval {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

arihant2math commented:
Can WPT be run on this? I think it builds, but my linker is malfunctioning at the moment.

sebsebmc commented:
You can also do a try run in your GitHub fork with `./mach try wpt -r [your remote]`

arihant2math commented on components/script/dom/idbopendbrequest.rs:
```diff
@@ -0,0 +1,371 @@
+        global
+            .database_access_task_source()
+            .queue(
+                task!(send_success_notification: move || {
```
Should this just be done outside of the database_task_source and in the dispatch_success function itself?

arihant2math commented on tests/wpt/meta/indexeddb/idb_webworkers.htm.ini:
```diff
@@ -0,0 +1,2 @@
+[idb_webworkers.htm]
+  expected: CRASH
```
I think fixing the crashes should be a followup PR.

arihant2math commented on components/script/dom/idbdatabase.rs:
```diff
@@ -0,0 +1,337 @@
+        // FIXME:(arihant2math) Remove from index set ...
+
+        // Step 7
+        // FIXME:(arihant2math!!!) Move this to constructor
```
Creating the IPC should probably be done in the constructor I think.

arihant2math commented:
@jdm I think this is good for another round of reviews. I think some error handling/IPC stuff needs cleanup but everything is resolved other than that.

arihant2math commented:
Also I've been looking through the spec, I see a bit of a contradiction:
```
  undefined close();
```
and https://www.w3.org/TR/IndexedDB-2/#closing-connection specifying that
> If the forced flag is set, then for each transaction [created](https://www.w3.org/TR/IndexedDB-2/#transaction-created) using connection run the steps to [abort a transaction](https://www.w3.org/TR/IndexedDB-2/#abort-a-transaction) with transaction and newly [created](https://www.w3.org/TR/WebIDL-1/#dfn-create-exception) "[AbortError](https://www.w3.org/TR/WebIDL-1/#aborterror)" [DOMException](https://www.w3.org/TR/WebIDL-1/#idl-DOMException).

I don't see how I can throw an error on a function that returns nothing

jdm commented:
Webidl doesn't have any special annotation for methods that can throw. Servo needs to add the `[Throws]` annotation to adjust the return type to allow returning exception values.

arihant2math commented:
bors try

arihant2math commented:
Whoops, I forgot what repo I'm on 🤣

arihant2math committed:
comment out unimplemented methods

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

arihant2math committed:
fix build

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

jdm commented:
I'm planning to re-review these changes in this coming week. I know this has been a slow process; I'm hopeful that subsequent PRs will be able to be incremental and easier to review!

jdm commented:
One warning that will need to be fixed:
```
error: unused variable: `db`
   --> components/net/indexeddb/idb_thread.rs:260:29
    |
260 |                 if let Some(db) = self.databases.remove(&idb_description) {
    |                             ^^ help: if this is intentional, prefix it with an underscore: `_db`
    |
```

arihant2math commented:
Great, thanks, I'll fix the warning shortly

jdm commented on components/script/dom/idbdatabase.rs:
```diff
@@ -0,0 +1,387 @@
+            .expect("Could not receive object store deletion status")
+            .is_err()
+        {
+            warn!("Object store deletion failed in idb thread");
```
```suggestion
            warn!("Object store deletion failed in idb thread");
            return Err(Error::InvalidState);
```

jdm commented on components/script/lib.rs:
```diff
@@ -81,6 +81,7 @@ mod xpath;
 mod svgpath;
 
 pub use init::init;
+pub(crate) use realms::enter_realm;
```
```suggestion
```

jdm commented on components/script/dom/idbopendbrequest.rs:
```diff
@@ -0,0 +1,371 @@
+use crate::dom::idbrequest::IDBRequest;
+use crate::dom::idbtransaction::IDBTransaction;
+use crate::dom::idbversionchangeevent::IDBVersionChangeEvent;
+use crate::enter_realm;
```
```suggestion
use crate::realms::enter_realm;
```

jdm commented on components/script/dom/idbopendbrequest.rs:
```diff
@@ -0,0 +1,371 @@
+        let this = Trusted::new(self);
+        let result = Trusted::new(result);
+
+        global.task_manager().database_access_task_source().queue(
```
```suggestion
        // TODO: some callers expect this to be synchronous, rather than an async task.
        global.task_manager().database_access_task_source().queue(
```

jdm commented on components/script/dom/idbrequest.rs:
```diff
@@ -0,0 +1,276 @@
+            );
+
+            transaction.set_active_flag(true);
+            event
+                .upcast::<Event>()
+                .fire(request.upcast(), CanGc::note());
```
```suggestion
            event.fire(request.upcast(), CanGc::note());
```

jdm commented on components/script/dom/idbrequest.rs:
```diff
@@ -0,0 +1,276 @@
+
+impl IDBRequestMethods<crate::DomTypeHolder> for IDBRequest {
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbrequest-result
+    fn Result(&self, _cx: SafeJSContext, _val: js::rust::MutableHandle<'_, js::jsapi::Value>) {
+        self.result.get();
```
```suggestion
    fn Result(&self, _cx: SafeJSContext, val: js::rust::MutableHandle<'_, js::jsapi::Value>) {
        val.set(self.result.get());
```

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -0,0 +1,321 @@
+#[dom_struct]
+pub struct IDBTransaction {
+    eventtarget: EventTarget,
+    object_store_names: DomRoot<DOMStringList>,
```
```suggestion
    object_store_names: Dom<DOMStringList>,
```

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -0,0 +1,320 @@
+    db: Dom<IDBDatabase>,
+    error: MutNullableDom<DOMException>,
+
+    // Not specified in WebIDL below this line
```
This comment seems incorrect now.

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -0,0 +1,321 @@
+    fn new_inherited(
+        connection: &IDBDatabase,
+        mode: IDBTransactionMode,
+        scope: DomRoot<DOMStringList>,
```
```suggestion
        scope: &DOMStringList,
```

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -0,0 +1,321 @@
+        global: &GlobalScope,
+        connection: &IDBDatabase,
+        mode: IDBTransactionMode,
+        scope: DomRoot<DOMStringList>,
```
```suggestion
        scope: &DOMStringList,
```

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -0,0 +1,321 @@
+
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbtransaction-objectstorenames
+    fn ObjectStoreNames(&self) -> DomRoot<DOMStringList> {
+        self.object_store_names.clone()
```
```suggestion
        self.object_store_names.as_rooted()
```

jdm reviewed (changes requested):
At this point I think it's going to be more useful to merge these changes and incrementally improve the code rather than going through additional large reviews. I didn't notice anything that made me want to prevent merging it at all, so let's address these comments and get it merged!

arihant2math committed:
apply review

Signed-off-by: Ashwin Naren <arihant2math@gmail.com>

jdm commented on components/script/dom/idbtransaction.rs:
```diff
@@ -80,7 +79,7 @@ impl IDBTransaction {
         connection: &IDBDatabase,
         mode: IDBTransactionMode,
-        scope: DomRoot<DOMStringList>,
+        scope: DOMStringList,
```
I think this is missing a &

jdm committed:
Fix tidy errors and revert unnecessary Cargo.lock changes.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm committed:
Update WPT results.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm commented:
I've gone ahead and fixed up the remaining build and test-tidy errors. Let's get this merged!

jdm committed:
Fix clippy warnings.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm commented:
The open Harmony error:
```
Error relocating /data/storage/el1/bundle/libs/arm64/libservoshell.so: pthread_mutexattr_setrobust: symbol not found
```
We'll need to figure out which dependency is bringing that in and what we can do to avoid it.

jschwe commented:
> Although it's confusing that trying the opemharmony job again succeeded...

No, it did fail. But currently failure is allowed for this job, so it doesn't block the MQ and sadly this also causes the bot to write it succeeded, instead of mentioning that there were errors, which are not required to pass the status checks.

jdm committed:
Disable pthread_mutexattr_setrobust in LMDB on OpenHarmony.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm committed:
Tidy.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>
